### PR TITLE
subsys/disk: add stm32 sdmmc disk access driver

### PR DESCRIPTION
**Pull request by anthonybrandon**
Add a disk access driver for the stm32 sdmmc component. The driver is
based around the stm32 cube HAL and uses the blocking API.

Signed-off-by: Anthony Brandon <anthony@amarulasolutions.com>

EDIT: Based on https://github.com/zephyrproject-rtos/zephyr/pull/24787 for fat_fs sample rework

### CODE REVIEW

**Review by erwango — CHANGES_REQUESTED** [review]
Thanks for providing this. I think this will interest many users.

Now, it is not possible to merge this change as is and will require some adaptations. Indeed, it is not possible to introduce as is a vendor specific subsystem, you need to provide support via a generic driver API.
Luckily, you can base your work on https://github.com/zephyrproject-rtos/zephyr/pull/13214 which was perfectly fine in that regard.

**Comment by anthonybrandon** [comment]
@erwango Can you give me some more guidance on how you would like me to change this?
Do you mean I shouldn't use the STM32 Cube HAL? I see other drivers that use it however, so I thought it would be okay.

**Comment by erwango** [comment]
> Do you mean I shouldn't use the STM32 Cube HAL? I see other drivers that use it however, so I thought it would be okay.

@anthonybrandon, you can, no problem with that. Issue here is that you're pushing a subsystem driver directly based on a stm32 peripheral. What I'd advice you to do is to implement your SDMMC disk driver based on the SDMM API proposed in #13214. Then STM32 HAL would be used in the SDMMC driver.

**Comment by anthonybrandon** [comment]
@erwango , okay, that makes sense to me. Looking at the sdmmc API  proposed in #13214 it seems a bit more low level than the one exposed in the stm32 cube HAL. So I don't think I can make it fit.

**Comment by erwango** [comment]
@anthonybrandon 

>  Looking at the sdmmc API proposed in #13214 it seems a bit more low level than the one exposed in the stm32 cube HAL. So I don't think I can make it fit.

Indeed. Then I see two options:
- Rebase and reuse sdmmc driver provided in #13214
- Propose an alternative SDMMC API, on top of your HAL based STM32 driver. And implement your disk access subsytem on top of it.

**Comment by anthonybrandon** [comment]
@erwango thanks. I'll give both options a try and see what works out better for me.

**Comment by anthonybrandon** [comment]
@erwango I've updated my change with a new API.
I've named it sdmmc_hal, since I intend it to be used for implementations based on whatever is provided by the vendor HAL (stm32 cube in this case, but should work for the esp hal as well).
If anyone ever implements a generic sdmmc driver, then this could be removed.
I looked into doing that, but I am not familiar enough with the low level sdmmc commands, and the references I looked at all were doing slightly different things, so I wasn't confident in that approach.

**Comment by galak** [comment]
dev-review (Mar 26): For now just implement this directly to the disk_access.h api and remove the middle sdmmc layer for now.  We can look at a sdmmc api once we see there are multiple users.

**Comment by anthonybrandon** [comment]
@galak Should I just revert to my original pull request in that case?

**Comment by erwango** [comment]
@anthonybrandon 

> @galak Should I just revert to my original pull request in that case?
Or am I misunderstanding what you are asking?

You got it right. Following dev-review meeting, it was decided the initial approach is ok as there is no others users for now for SDMMC api. Sorry for the back and forth change.

**Comment by anthonybrandon** [comment]
@erwango Sorry, been busy recently. I need to rebase and retest the original patch. I'll try to have it done soon.

File: subsys/disk/Kconfig
Comment (erwango) [inline subsys/disk/Kconfig:139]: This should not be needed if you're using only the .h part of the ll api.
Can you check this is really needed (since I haven't seen use of ll .c function in the driver below) ?

File: subsys/disk/disk_access_stm32_sdmmc.c
Comment (erwango) [inline subsys/disk/disk_access_stm32_sdmmc.c:6]: Please put following line here:
```
#define DT_DRV_COMPAT st_stm32_sdmmc
```
Then use DT macros that don't need compatible argument

File: subsys/disk/Kconfig
Comment (erwango) [inline subsys/disk/Kconfig:140]: Please add default value based on device tree node presence:
```
	default $(dt_compat_enabled,$(DT_COMPAT_ST_STM32_SDMMC))
```
With following defined upper in the same file:
```
DT_COMPAT_ST_STM32_SDMMC := st,stm32-sdmmc
```

File: subsys/disk/disk_access_stm32_sdmmc.c
Comment (erwango) [inline subsys/disk/disk_access_stm32_sdmmc.c:369]: This part should not be needed with the Kconfig tweak proposed above.

**Review by erwango — CHANGES_REQUESTED** [review]
Few comments, otherwise it seems good, but I'm adding some people to the review to cross-check.

One request though: we need the driver to be compiled as part of sanitycheck/CI.
Can you propose a sample/test that could make use of it (along with a board config)?

**Comment by erwango** [comment]
^^@ABOSTM

**Comment by anthonybrandon** [comment]
@erwango thanks for the review. I will try to address all your comments.
Regarding the codeowners build failure, who should I put as codeowner?

File: subsys/disk/Kconfig
Comment (anthonybrandon) [reply subsys/disk/Kconfig:139]: Without it I get link errors like:
```
modules/hal/stm32/stm32cube/stm32l4xx/drivers/src/stm32l4xx_hal_sd.c:947: undefined reference to `SDMMC_ConfigData'
```
Maybe USE_STM32_HAL_SD should depend on USE_STM32_LL_SDMMC?

**Comment by erwango** [comment]
> Regarding the codeowners build failure, who should I put as codeowner?

Yourself :-)

**Comment by anthonybrandon** [comment]
@erwango I think I've addressed all your previous comments.
Regarding the sanitycheck test, I've added one that only performs build.
I selected the stm32l496g_disco board which should have an sdmmc slot.
I don't have this board, so I can't really check, but it's only used for building so I don't think it should matter. The gpio's I used in the overlay are not correct. The stm32l496g_disco has some expansion gpios that are used and as far as I can tell those are not supported. The stm32l496g_disco also has no pwr-gpio. So maybe that one should be optional.

File: CODEOWNERS
Comment (erwango) [inline CODEOWNERS:481]: /tests/subsys/disk/stm32_sdmmc/

File: subsys/disk/Kconfig
Comment (erwango) [reply subsys/disk/Kconfig:141]: ok, let's keep it as is

File: tests/subsys/disk/stm32_sdmmc/boards/stm32l496g_disco.overlay
Comment (erwango) [inline tests/subsys/disk/stm32_sdmmc/boards/stm32l496g_disco.overlay:16]: Remove empty line

File: tests/subsys/disk/stm32_sdmmc/prj.conf
Comment (erwango) [inline tests/subsys/disk/stm32_sdmmc/prj.conf:1]: tests are expected to provide normalized test verdict.
This should rather be moved to the sample section: samples/subsys

File: tests/subsys/disk/stm32_sdmmc/src/main.c
Comment (erwango) [inline tests/subsys/disk/stm32_sdmmc/src/main.c:14]: Thanks for proposing this sample.
Though, can you provide an even quite basic example on how to exercise APIs?

**Review by erwango — CHANGES_REQUESTED** [review]
For board support, can you also check stm32f469i_disco?

File: tests/subsys/disk/stm32_sdmmc/src/main.c
Comment (anthonybrandon) [reply tests/subsys/disk/stm32_sdmmc/src/main.c:14]: For my project I just use it through the file system API. Would it be fine to enable FAT and then in the sample just mount it and write something?
Or should I directly use the disk API?

**Comment by anthonybrandon** [comment]
@erwango 
> For board support, can you also check stm32f469i_disco?
I looked into it. It seems the low level defines are a bit different, and probably the clock configuration as well. Some of it can be handled by simply using some ifdef to select between SDIO_ and SDMMC_, but the clock enable part would be a bit more complex to ensure a good frequency is selected. As far as I know zephyr does not (yet) handle the PLLSAI clocks for stm32. So the low level part is done in this driver instead. That will differ for various SoC.
I'm not sure what the right way forward is.

File: tests/subsys/disk/stm32_sdmmc/src/main.c
Comment (erwango) [reply tests/subsys/disk/stm32_sdmmc/src/main.c:14]: > Would it be fine to enable FAT and then in the sample just mount it and write something?

@anthonybrandon would you manage `samples/subsys/fs/fat_fs` compiling/running on that board ?

**Comment by erwango** [comment]
> @erwango
> 
> > For board support, can you also check stm32f469i_disco?
> 
> I looked into it. It seems the low level defines are a bit different, and probably the clock configuration as well. Some of it can be handled by simply using some ifdef to select between SDIO_ and SDMMC_, but the clock enable part would be a bit more complex to ensure a good frequency is selected. As far as I know zephyr does not (yet) handle the PLLSAI clocks for stm32. So the low level part is done in this driver instead. That will differ for various SoC.
> I'm not sure what the right way forward is.

If not too much trouble, we can afford a few #ifdef to allow support on some SoCs, and configuration can be provided using Kconfig. If it allows to have a sample working on a board available upstream, this is worth the pain.

File: dts/bindings/mmc/st,stm32-sdmmc.yaml
Comment (hjuul) [inline dts/bindings/mmc/st,stm32-sdmmc.yaml:20]: What does `pwr-gpios` do? Is it something custom for a particular board? 
If so I suggest you keep it out of the sdmmc binding, I suppose most implementations won't use a power pin with SD cards.

File: tests/subsys/disk/stm32_sdmmc/boards/stm32l496g_disco.overlay
Comment (hjuul) [inline tests/subsys/disk/stm32_sdmmc/boards/stm32l496g_disco.overlay:13]: I'm not familiar with stm32l496g_disco, but a brief look at the schematics suggest that `PA15` is not the correct value for `cd-gpios`.

**Comment by hjuul** [comment]
> > @erwango
> > > For board support, can you also check stm32f469i_disco?
> > 
> > 
> > I looked into it. It seems the low level defines are a bit different, and probably the clock configuration as well. Some of it can be handled by simply using some ifdef to select between SDIO_ and SDMMC_, but the clock enable part would be a bit more complex to ensure a good frequency is selected. As far as I know zephyr does not (yet) handle the PLLSAI clocks for stm32. So the low level part is done in this driver instead. That will differ for various SoC.
> > I'm not sure what the right way forward is.
> 
> If not too much trouble, we can afford a few #ifdef to allow support on some SoCs, and configuration can be provided using Kconfig. If it allows to have a sample working on a board available upstream, this is worth the pain.

@anthonybrandon is it at all possible to use STM32Cube HAL instead of LL when configuring clocks? That should make the code compatible across stm32 families. I'm trying to get this to work for stm32f746g_disco and can help verify the implementation on that platform.

File: dts/bindings/mmc/st,stm32-sdmmc.yaml
Comment (anthonybrandon) [reply dts/bindings/mmc/st,stm32-sdmmc.yaml:20]: Yes, it is for my board. I can remove it from the driver.

File: tests/subsys/disk/stm32_sdmmc/boards/stm32l496g_disco.overlay
Comment (anthonybrandon) [reply tests/subsys/disk/stm32_sdmmc/boards/stm32l496g_disco.overlay:13]: I think I mentioned this in one of my earlier comments, but as far as I can tell the gpio it's connected to is on some peripheral that is not currently supported. I just picked the stm32l496g_disco because board is similar to what I have and it has sdcard. I don't actually have that board. I could make the card detect pin optional and have the driver assume the sdcard is present if the gpio is not set.

**Comment by anthonybrandon** [comment]
> 
> @anthonybrandon is it at all possible to use STM32Cube HAL instead of LL when configuring clocks? That should make the code compatible across stm32 families. I'm trying to get this to work for stm32f746g_disco and can help verify the implementation on that platform.

I'm not sure. I have to take a close look at what is available in the HAL. Looking at projects generated by the stm32cube tools the clock setup is done once at the start using `HAL_RCC_ClockConfig` which also sets the clocks for other peripherals. I can take a look and see if that is usable for setting the clock for just one peripheral.

**Comment by anthonybrandon** [comment]
@hjuul As far as I can tell the HAL is different for L4 and F4 (and possibly other) SoCs specifically around the issue of sdmmc/sdio and the clock tree.
So even if it's possible to use HAL instead of LL there will need to be ifdefs.

**Comment by hjuul** [comment]
> @hjuul As far as I can tell the HAL is different for L4 and F4 (and possibly other) SoCs specifically around the issue of sdmmc/sdio and the clock tree.
> So even if it's possible to use HAL instead of LL there will need to be ifdefs.

Ok, for stm32f746g_disco it works if I just remove the LL calls in `stm32_sdmmc_clock_enable()` (currently lines 43-59). This is because it uses the main PLL (not PLLSAI) as input to PLL48CLK by default.
So #ifdefs will work, the challenge is to use the right conditions (could be Kconfig).

File: tests/subsys/disk/stm32_sdmmc/boards/stm32l496g_disco.overlay
Comment (hjuul) [inline tests/subsys/disk/stm32_sdmmc/boards/stm32l496g_disco.overlay:8]: Should we also add sdmmc1 entries to the `.dtsi` files? That would make it easy to add support for sdmmc1 in board `.dts` files (don't need overlay files).

**Comment by hjuul** [comment]
@anthonybrandon I've managed to get your code working on stm32f746_disco board. I mostly had to change board- and family specific files. If it is ok with you I'd like to add support for this board as part of this PR.
What is the best way for me to contribute the code? Can I add to this PR or should I send you a patch?

On another matter: Have you considered using the SDMMC with DMA? I will probably add that for my project, but perhaps that's a topic for a different  PR..

**Comment by anthonybrandon** [comment]
> @anthonybrandon I've managed to get your code working on stm32f746_disco board. I mostly had to change board- and family specific files. If it is ok with you I'd like to add support for this board as part of this PR.
> What is the best way for me to contribute the code? Can I add to this PR or should I send you a patch?
> 

I'm not very familiar with githubs workflow. If there is a way for you to add to this PR I don't mind. Otherwise send me the patch and I will add it to this PR. If you upload your change to your own github I can pull it from there.

> On another matter: Have you considered using the SDMMC with DMA? I will probably add that for my project, but perhaps that's a topic for a different PR..

I considered it, but I think that is something for a separate PR.

File: tests/subsys/disk/stm32_sdmmc/boards/stm32l496g_disco.overlay
Comment (anthonybrandon) [reply tests/subsys/disk/stm32_sdmmc/boards/stm32l496g_disco.overlay:8]: Yes, if it's tested on the board I think it would probably be right to add support in the board dts instead of overlay.

File: dts/bindings/mmc/st,stm32-sdmmc.yaml
Comment (erwango) [reply dts/bindings/mmc/st,stm32-sdmmc.yaml:20]: I think it is ok to make it optional: `required: false`

File: tests/subsys/disk/stm32_sdmmc/boards/stm32l496g_disco.overlay
Comment (erwango) [reply tests/subsys/disk/stm32_sdmmc/boards/stm32l496g_disco.overlay:8]: +1 to move things to .dtsi/.dts when possible.
A node should be defined in .dtsi (undet dts/arm/st/...), then edited in board .dts for customization.
Overlay files should be used for application specific purpose when needed

**Comment by erwango** [comment]
@hjuul Thanks for your help, this will help converging to a upstream ready solution.
A suggestion for contributing your code is to push a new PR that includes commits from this PR, so we can see the additions that are needed to work on stm32f746g_disco.
Btw, is there on board SD/MMC on this board?

**Comment by anthonybrandon** [comment]
@erwango @hjuul 

- I've made both cd-gpios and pwr-gpios optional. If no card detect is specified it assumes there is a card. The reason I did this is becuase the stm32l496g_disco board has a card detect pin on an expansion device which is not supported as far as I can tell. 
- I've removed the test specific to this driver.
- I've updated the stm32l471.dtsi to add the device node. I've updated the stm32l496 board to enable it and have correct pinmuxing. 
- I've updated the fat_fs sample to compile for this board.

I'm looking forward to Hjuul's patch for stm32f7.

**Comment by hjuul** [comment]
> @hjuul Thanks for your help, this will help converging to a upstream ready solution.
> A suggestion for contributing your code is to push a new PR that includes commits from this PR, so we can see the additions that are needed to work on stm32f746g_disco.

@erwango  I've forked Zephyr, merged in the commits from this PR, committed my changes and pushed to https://github.com/hjuul/zephyr/tree/stm32f746g_disco-sdmmc.
Should I create a PR to `anthonybrandon:stm32-sdmmc` or to `zephyr:master` ?

> Btw, is there on board SD/MMC on this board?

Yes

**Comment by anthonybrandon** [comment]
> @erwango I've forked Zephyr, merged in the commits from this PR, committed my changes and pushed to https://github.com/hjuul/zephyr/tree/stm32f746g_disco-sdmmc.
> Should I create a PR to `anthonybrandon:stm32-sdmmc` or to `zephyr:master` ?
I'll pick it from there and add it here.

**Comment by anthonybrandon** [comment]
@hjuul I've updated the my patch series. Please confirm that it still works for you.
The base address is now taken from the device tree instead of the HAL macro SDMMC1, which does not exist for every SoC series and is more consistent with how other devices are handled in the dts.
I also made a small change to the pinmux in your patch to use the DT_HAS_NODE macro.

File: dts/arm/st/l4/stm32l471.dtsi
Comment (hjuul) [inline dts/arm/st/l4/stm32l471.dtsi:197]: Minor detail: It seems everywhere else there is an underscore before the 1. So `SDMMC1` -> `SDMMC_1`.

**Comment by hjuul** [comment]
> @hjuul I've updated the my patch series. Please confirm that it still works for you.


I just checked with your latest changes, it still works for stm32f746g_disco.

> I also made a small change to the pinmux in your patch to use the DT_HAS_NODE macro.

Yeah, I chose to use Kconfig ifdef instead. Honestly I don't know what's the best choice here. @erwango , any opinion?

File: dts/arm/st/l4/stm32l471.dtsi
Comment (anthonybrandon) [reply dts/arm/st/l4/stm32l471.dtsi]: Fixed, thanks.

**Comment by erwango** [comment]
> Yeah, I chose to use Kconfig ifdef instead. Honestly I don't know what's the best choice here. @erwango , any opinion?

We're moving toward using DT api whenever is possible,so DT_HAS_NODE is the good choice here.
Seems there are been good progress globally. I'll review in detail on monday.

File: subsys/disk/disk_access_stm32_sdmmc.c
Comment (erwango) [inline subsys/disk/disk_access_stm32_sdmmc.c:126]: Can you check that LOG_DBG is not already returning function name ?

File: samples/subsys/fs/fat_fs/prj_stm32f746g_disco.conf
Comment (erwango) [inline samples/subsys/fs/fat_fs/prj_stm32f746g_disco.conf:5]: IIC These are all already available in prj.conf, so this file should'nt be needed

File: samples/subsys/fs/fat_fs/prj_stm32l496g_disco.conf
Comment (erwango) [inline samples/subsys/fs/fat_fs/prj_stm32l496g_disco.conf:5]: IIC These are all already available in prj.conf, so this file should'nt be needed

**Review by erwango — CHANGES_REQUESTED** [review]
LGTM, 2 requests though:
-I think you can squash the 3 first commits.
-Also, can you update sample's README ?

File: subsys/disk/disk_access_stm32_sdmmc.c
Comment (anthonybrandon) [reply subsys/disk/disk_access_stm32_sdmmc.c:126]: I think it does. Should I just remove all the DBG messages?

File: samples/subsys/fs/fat_fs/prj_stm32f746g_disco.conf
Comment (anthonybrandon) [reply samples/subsys/fs/fat_fs/prj_stm32f746g_disco.conf:5]: The default prj.conf for that sample has:
`CONFIG_DISK_ACCESS_SPI_SDHC=y`
Which causes a build failure because the stm32 boards don't have that.

**Comment by anthonybrandon** [comment]
@erwango I've squashed the three first patches, and removed the debug prints. I've added a separate commit to update the fat_fs sample readme.

File: samples/subsys/fs/fat_fs/prj_stm32l496g_disco.conf
Comment (MaureenHelm) [reply samples/subsys/fs/fat_fs/prj_stm32l496g_disco.conf:5]: This conf file isn't under `boards/`, so it gets used instead of `prj.conf`. I proposed a refactoring of this sample's conf files in #24787

File: samples/subsys/fs/fat_fs/sample.yaml
Comment (MaureenHelm) [inline samples/subsys/fs/fat_fs/sample.yaml]: I proposed a refactoring of the sample yaml in #24787 to reduce the platform whitelisting. If merged, it means you would instead add `sdhc` to the supported list in your board yaml.

File: samples/subsys/fs/fat_fs/prj_stm32l496g_disco.conf
Comment (erwango) [reply samples/subsys/fs/fat_fs/prj_stm32l496g_disco.conf:5]: This seems cleaner indeed, txs @MaureenHelm.
@anthonybrandon can you use https://github.com/zephyrproject-rtos/zephyr/pull/24787/commits/3649c16046c083e2f6d3951791cc4916180901d4 and do you changes on top?

File: samples/subsys/fs/fat_fs/sample.yaml
Comment (anthonybrandon) [reply samples/subsys/fs/fat_fs/sample.yaml]: Does that mean I should move `DISK_ACCESS_STM32_SDMMC` under `DISK_ACCESS_SDHC` in the Kconfig?

Comment (MaureenHelm) [reply samples/subsys/fs/fat_fs/sample.yaml]: That's a good point, I'm not sure if it makes sense to move it. Keeping it there means we'll either need `samples/subsys/fs/fat_fs/prj_stm32l496g_disco.conf` like you originally proposed, or move `CONFIG_DISK_ACCESS_SDHC=y` to board conf overlay files in `samples/subsys/fs/fat_fs/boards/`. I'm ok with either option.

@erwango what do you think?

Comment (anthonybrandon) [reply samples/subsys/fs/fat_fs/sample.yaml]: @MaureenHelm I checked and building with `CONFIG_DISK_ACCESS_SDHC=y` does work for the STM32 boards, since all it seems to do is enable the submenu in Kconfig. So it could just stay as is. It's just superfluous for the stm32_sdmmc driver.
Not having the SDHC_SPI enabled is enough to allow it to build. So from that standpoint it would be okay to leave as is.

Comment (erwango) [reply samples/subsys/fs/fat_fs/sample.yaml]: Yes, let's drop the readme.
For the DISK_ACCESS_SDHC, I'm not so sure, I don't think you need it enabled in your case, so it would be part of each board configuration.

File: samples/subsys/fs/fat_fs/prj_stm32f746g_disco.conf
Comment (erwango) [reply samples/subsys/fs/fat_fs/prj_stm32f746g_disco.conf]: As I commented in #24787, CONFIG_DISK_ACCESS_SPI_SDHC=y might be better put in each boards .conf so you can have this board to be treated like other boards under /boards

**Comment by hjuul** [comment]
@anthonybrandon I'm very eager to get these changes into master branch before the merge window close for release 2.3, which is Friday this week (May 8th).
Please let me know if there is anything I can do to help you speed things up.

File: samples/subsys/fs/fat_fs/prj_stm32f746g_disco.conf
Comment (erwango) [reply samples/subsys/fs/fat_fs/prj_stm32f746g_disco.conf]: @anthonybrandon https://github.com/zephyrproject-rtos/zephyr/pull/24787 is now taking this into account.
You just need to rebase and use same config scheme

**Comment by anthonybrandon** [comment]
@hjuul other than maybe retesting that your board still works (I didn't change anything major so I don't see why it wouldn't) I don't think there is anything left to do at the moment.

File: samples/subsys/fs/fat_fs/prj_stm32f746g_disco.conf
Comment (anthonybrandon) [reply samples/subsys/fs/fat_fs/prj_stm32f746g_disco.conf]: okay, done.

**Comment by hjuul** [comment]
> 
> 
> @hjuul other than maybe retesting that your board still works (I didn't change anything major so I don't see why it wouldn't) I don't think there is anything left to do at the moment.

Great. I just checked again with my board (stm32f746g_disco) using the latest commit in this PR, and it still works.
Thank you very much for providing this work @anthonybrandon, I really appreciate it!

@erwango , do you think it is realistic to get this PR into master before the feature merge window this friday?

File: samples/subsys/fs/fat_fs/prj_stm32f746g_disco.conf
Comment (erwango) [reply samples/subsys/fs/fat_fs/prj_stm32f746g_disco.conf]: @anthonybrandon This is not exactly what I expect. Sorry if not clear. The following are already available in prj.cnf:
```
CONFIG_DISK_ACCESS=y
CONFIG_LOG=y
CONFIG_FILE_SYSTEM=y
CONFIG_FAT_FILESYSTEM_ELM=y
```
So, if you really need printk, you just need to add a file `samples/subsys/fs/fat_fs/boards/stm32f746g_disco.conf` with the following:
```
CONFIG_PRINTK=y
```

**Review by erwango — CHANGES_REQUESTED** [review]
@anthonybrandon not sure if this is an oversight or not, but stm32f746g_disco commit still requires a change

**Comment by anthonybrandon** [comment]
@erwango yes, I thought I had removed that, but apparently I forgot.

**Review by erwango — APPROVED** [review]

**Comment by erwango** [comment]
> @erwango , do you think it is realistic to get this PR into master before the feature merge window this friday?

@hjuul Hopefully yes. Btw, if you're satifsfied, approving the PR & mention in the review that this is working on your side will help

**Review by hjuul — APPROVED** [review]
I have verified that this is running fine on the stm32f746_disco board.

File: boards/arm/stm32l496g_disco/pinmux.c
Comment (erwango) [inline boards/arm/stm32l496g_disco/pinmux.c:42]: Replace `DT_HAS_NODE` with `DT_HAS_NODE_STATUS_OKAY`

File: boards/arm/stm32f746g_disco/pinmux.c
Comment (erwango) [inline boards/arm/stm32f746g_disco/pinmux.c:61]: Replace `DT_HAS_NODE` with `DT_HAS_NODE_STATUS_OKAY`

**Review by erwango — CHANGES_REQUESTED** [review]
@anthonybrandon can you rebase on master and apply following changes?

File: boards/arm/stm32l496g_disco/pinmux.c
Comment (anthonybrandon) [reply boards/arm/stm32l496g_disco/pinmux.c]: Done

**Review by erwango — APPROVED** [review]
Txs!

**Comment by erwango** [comment]
@anthonybrandon , since #24787 has been merged, you can try to rebase.
This is not mandatory, but with some luck it will also make CI pass.

File: subsys/disk/disk_access_stm32_sdmmc.c
Comment (erwango) [inline subsys/disk/disk_access_stm32_sdmmc.c:390]: Can you try #if DT_HAS_NODE_STATUS_OKAY(DT_DRV_INST(0))?

**Review by erwango — CHANGES_REQUESTED** [review]
DT macro to be updated.
@anthonybrandon we're in live DT API update, some more changes are needed.
Thanks for bearing with us.

File: subsys/disk/disk_access_stm32_sdmmc.c
Comment (anthonybrandon) [reply subsys/disk/disk_access_stm32_sdmmc.c]: Sorry. Didn't build it myself after rebasing.

**Review by erwango — APPROVED** [review]

File: dts/bindings/mmc/st,stm32-sdmmc.yaml
Comment (galak) [inline dts/bindings/mmc/st,stm32-sdmmc.yaml:11]: shouldn't need `type: string` here.

**Comment by carlescufi** [comment]
Merging pending small tweak to the binding, which can be provided later.

File: dts/bindings/mmc/st,stm32-sdmmc.yaml
Comment (anthonybrandon) [reply dts/bindings/mmc/st,stm32-sdmmc.yaml:11]: I assume I should make a new pull request for that? Or can I still update this one after it's merged?

Comment (anthonybrandon) [reply dts/bindings/mmc/st,stm32-sdmmc.yaml:11]: Fixed in #25132

**Pull request merged**
